New post: "AWS IoT, Elixir, and Nerves: A Crash Course" (#1)
danielspofford/blogposts: posts/aws-iot-elixir-nerves.md

# AWS IoT Core, Elixir, and Nerves: A Crash Course

Find out what is happening at the intersection of an IoT device management and

> pubsub service, and a fault tolerant language leveraging a powerful platform.

Reviewer: I share a similar sentiment as @squaresurf -- but I'm not sure that I can offer anything better... It _is_ an accurate description.

> pubsub service, and a fault tolerant language leveraging a powerful platform.

Author: ```
Build on a solid foundation of Nerves, quickly write robust firmware in Elixir,
and rocket through the clouds with AWS. Along the way this post will expose you
to some of the biggest pieces of the IoT puzzle.
```

--

## Introductions

AWS IoT Core is a managed cloud service that facilitates managing devices,
securely communicating to and from them, and taking action based on their
messages.

> Elixir is a functional language well suited for maintainable, low-latency, and

Reviewer: ... language `that` is well suited ...

Elixir".

NervesHub facilitates secure and dynamic OTA firmware updates for embedded
devices. It is a managed service but it can be hosted privately, and there are
a variety of other Elixir libraries under the NervesHub namespace.

## General Management

### AWS IoT Core

AWS IoT Core manages many resources fundamental to operating an IoT deployment.
These include but are not limited to: Things, Types, Thing Groups, Billing
Groups, and Jobs. A Thing is the AWS term for some IoT device, it could be a
doorbell, a camera, a thermometer, or in reality, anything. These Things allow
you to manage and take certain actions even when a Thing is not currently
online. For example, Device Shadows allow you to read the last known state of a
Thing and even edit it, so that the next time that Thing comes online it could
respect those changes.

Types are just categories of Things, Thing Groups allow you to attach policies
and configure logging options at the Group instead of Thing level. Billing
Groups are similar to Thing Groups, but can be associated with tags which
facilitates categorization and tracking of IoT related costs. Finally, Jobs are

> essentially a remote command that will be interpretted by one or many Things,

Reviewer: ~interpretted~ interpreted

can be assigned tags. A Firmware is a signed chunk of binary information
associated with some metadata like version, VCS identifier, author,
time-to-live, etc.

Deployments are where the rubber meets the road. They represent a resource that
can be assigned a Firmware, tags, and version requirements among other things.
This facilitates very powerful resolution of eligible firmwares for a Device.
Through version requirements and tags one can dictate a complex multi-step
upgrade path for Devices.

## Communication

### AWS IoT Core

The communication protocol of choice between Things and AWS IoT Core is MQTT due
to its low bandwidth overhead, small code footprint on clients, and topic based
messaging. It is worth noting that while MQTT has 3 QoS levels (0, 1, 2), AWS
IoT only supports 0 and 1. Remember that Things can not only publish messsages
to topics to be processed in the cloud, but also can subscribe to topics so that
information can be pushed down to them as well.

To tie off that functionality are Actions: a convenient way to react to messages
published to topics. This is not an inflexible configuration either, AWS IoT

> Core allows you to leverage MQTT topic wild cards like `+` and `#` to specify

Reviewer: Might be good to link to the official docs on these wildcards and other topic specifications.

multiple topics at once. When an action is triggered you can take predefined
actions like writing to a dynamo table, adjusting a CloudWatch alarm, or a whole
host of other actions, but the escape hatch is invoking a lambda which allows
you to do whatever you want.

### NervesHub

NervesHub and AWS IoT do not need to talk to each other, but your device does
need to talk to NervesHub. For this you can leverage the Elixir library

> aptly named `NervesHub`. This will allow your device to maintain a websocket

Reviewer: Link out to hex

Author: I'm going to do GitHub, but yes, links are tight.

connection to the NervesHub service to allow real-time pushes of firmware and
facilitate useful metadata like whether a device is currently online or not.

This always-on connection may be restrictive contextually if you have strict
power or bandwidth concerns. In which case you simply do not leverage the
real-time features of the NervesHub service and instead opt to poll at your
leisure. In order for your device running Elixir to speak MQTT to AWS IoT, Martin

> Gausby's Tortoise library will do the trick.

Reviewer: Link to tortoise

## Security

> ### AWS IoT Core

Reviewer: Check this out -- https://docs.aws.amazon.com/iot/latest/developerguide/iot-security-identity.html

It feels like this section is not super clear that there are two distinct problems (though you cover both steps in the bullet points below)
* Identity Verification
* Transport Security

You can actually leverage other methods besides X.509 for identity verification if that is desired (I'm not sure it's a good idea though).

Author: Reworked the section a bit.

The `partial_chain/1` function resolves successfully against Amazon root CAs.
Keep in mind when connecting to AWS IoT to:

  * provide the device certificate and key.
  * Provide a set of CA certificates including:
    * Amazon's root CA(s).
    * the CA who signed the device certificate which may or may not be Amazon's
      root CA.

All of that done successfully, a device may connect and publish/subscribe to its
heart's content. Or at least until it runs into an AWS IoT Policy saying
otherwise.

### NervesHub

NervesHub leverages the same security story as AWS IoT: public key cryptography
to ensure secure communications to and from Devices. It also has a similar
pattern of allowing you to generate device certificates and keys using a
NervesHub CA. This certificate and key pair should be the same pair as is
recognized by AWS IoT. When using the NervesHub library to talk to the NervesHub
service it leverages the fact that the device is running on the Nerves platform
to know where the device certificate and key are, similarly if they were
generated by a NervesHub CA, that CA cert will automatically be sent along with
any requests.

## Just-in-Time Provisioning

### AWS IoT Core

> Recall the security section above, the manual steps I mentioned can be

Reviewer: Might be good to link to some AWS docs about this.

# AWS IoT Core, Elixir, and Nerves: A Crash Course

Find out what is happening at the intersection of an IoT device management and
pubsub service, and a fault tolerant language leveraging a powerful platform.

--

## Introductions

AWS IoT Core is a managed cloud service that facilitates managing devices,
securely communicating to and from them, and taking action based on their
messages.

Elixir is a functional language well suited for maintainable, low-latency, and
fault-tolerant systems. It runs on the Erlang VM which allows it to leverage
decades of production battle-testing and libraries.

Nerves is the combination of a platform, a framework, and tooling, where the
output is the ability to "craft and deploy bulletproof embedded software in
Elixir".

NervesHub is written in Elixir by the Nerves core team, though both are open

> source and contributed to by many others as well. It facilitate secure and

Author: ```suggestion
source and contributed to by many others as well. It facilitates secure and
```

dynamic OTA firmware updates for embedded devices. Keep in mind it is a managed

> service, but it can be managed privately as well, and there is a host of

Author: ```suggestion
service, but it can be hosted privately as well, and there is a variety of
```

Thing starting, working on, and finishing a Job, it will report its progress to
AWS IoT.

### NervesHub

NervesHub manages the CRUD of many important pieces of the secure over-the-air
(OTA) firmware update puzzle including but not limited to: Products, Devices,
Firmwares, and Deployments. At the time of writing, a Product is a name that
acts as grouping for Devices, Firmwares, and Deployments. A Device in NervesHub
is conceptually analagous to a Thing in AWS IoT. It has a name, is associated
with a device key and certificate (though NervesHub is only ever privy to the
certificate), and it can be assigned tags. A Firmware is a signed chunk of
binary information associated with some metadata like version, VCS identifier,
author, time-to-live, etc.

Deployments are where the rubber meets the road. They represent a resource that
can be assigned a Firmware, tags, and version requirements among other things.
This facilitates very powerful resolution of eligible firmwares for a Device.
Through version requirements and tags one can dictate a complex multi-step
upgrade path for Devices.

## Communication

### AWS IoT Core

> The communication protocol of choice between Things and Iot AWS Core is MQTT due

Author: ```suggestion
The communication protocol of choice between Things and AWS IoT Core is MQTT due
```
